### ISSUE TITLE
Upload / GetFileSize problem with SFTP

### ISSUE THREAD

**Issue by st4ck** [FTPbox/FTPbox#57]
Using SFTP (tested 5 different servers), FTPBox doesn't work as expected.
After the upload, FTPBox is not able to determine the file size and to rename it
After a long debug I've solved but I've not time to fork and make a pull request, so I write here the solution:

In FTPBoxLib / Client.cs in the function SafeUpload add the row with "++"

               return TransferStatus.Failure;
            }

            ++if (FTP) CheckWorkingDirectory();

            if (i.Item.Size == SizeOf(temp))
            {
                if (Exists(i.CommonPath)) Remove(i.CommonPath);
                Rename(temp, i.CommonPath);
If Remote directory was different from / the path was wrong. This solve the problem

Also be sure that FTP transfer is BINARY. Otherwise SIZE will fail.
I've solved adding in the FluentFTP sources and recompiling the library and changing the funcion GetFileSize (m_lock block) in FtpClient_Management.cs in:

                var savedDataType = CurrentDataType;
                SetDataTypeInternal(FtpDataType.Binary);
                if (!(reply = Execute("SIZE " + path.GetFtpPath())).Success)
					length = -1;
				else if (!long.TryParse(reply.Message, out length))
					length = -1;

                this.SetDataTypeInternal(savedDataType);
Probably this problem can be solved from FTPBox without touching library. This was the fastest solution for me

Best regards

**Comment by JohnTheGr8** [comment]
You said you use SFTP but the line you added does nothing for SFTP. Did you mean you're connecting with FTPS? If so, can you tell me what server you're running so I can try to reproduce the problem?

Also, I have pushed some changes to the code just recently. If you can test with the new code and tell me how it works, it would be helpful.

Regards

**Comment by st4ck** [comment]
Yes FTPS. Latest version of Proftpd with TLSv1.
I've the snapshot of 17 october but I'll try to update my code with latest changes (but I think doesn't affect my problem)
Anyway I've other problems that I've fixed like in function SafeUpload(SyncQueueItem i)
When checking if (i.Item.Size == SizeOf(temp)) in the majority of cases, the SIZE returned was wrong (smallest that the item size); but if I wait something like 500ms the SIZE returned will be correct.
I don't know if depends on async upload, but for now I've solved adding a Thread.Sleep before this check.
Also there is a bug in FluentFTP when changing to Binary type: if upload fail, the exists return "No such file or directory" after some while and when changing type the reply contains "No such file or directory" and not the correct response to TYPE I (causing crashing). So I've added a while in SetDataTypeInternal of FluentFTP
```
case FtpDataType.Binary:
while (!(reply = Execute("TYPE I")).Success)
{
    if (reply.Code != "550")
    {
       throw new FtpCommandException(reply);
     }
}
```

**Comment by st4ck** [comment]
Update: latest snapshot doesn't work with TLSv1. Doesn't connect at all with the server

**Comment by JohnTheGr8** [comment]
```
Yes FTPS. Latest version of Proftpd with TLSv1.
```

I am testing on ProFTPd with TLSv1.2 and FTPbox uses the TLSv1.1 and TLSv1.2 flags, that explains it. I'll fix this.

```
Anyway I've other problems that I've fixed like [...]
```

I understand that these issues are related to your modified code, in which you have switched the FTP library to FluentFTP, correct? I hesitate to switch to FluentFTP because from my tests it isn't very stable.

**Comment by st4ck** [comment]
I was this problems before switching to FluentFTP, but probably was a problem of previous snapshot of FTPbox. Now I'll compile your latest snapshot and I'll try with TLSv1.2

**Comment by JohnTheGr8** [comment]
Updated to add the TLSv1 flag, it should now connect. Please try the latest code as it is with TLSv1.

Thank you

**Comment by st4ck** [comment]
Ok seems connect now. But I started to have same problems.
The first one is when SafeUpload check file exists or not
Here
```
            if (Controller.FileLog.IsEmpty())
            {
                // TODO: allow user to select if the following should happen
                // skip synchronization if the file already exists and has the exact same size
                if (Exists(i.CommonPath) && SizeOf(i.CommonPath) == i.Item.Size)
                {
                    Log.Write(l.Client, "File seems to be already synced (skipping): {0}", i.CommonPath);
                    return TransferStatus.Success;
                }
            }
```

There is a logic problem. This block check if the file exists only if the file log is empty? So the first one file (if exists) will be skipped, next files will be uploaded event if exists online and have the same size.
I removed in my previous code if (Controller.FileLog.IsEmpty()) to check every file

**Comment by st4ck** [comment]
Also doesn't save my profile after exit. Every timeI've to reconfigure the account, so I think saving is not correctly implemented yet

**Comment by JohnTheGr8** [comment]
> There is a logic problem. This block check if the file exists only if the file log is empty? So the first one file (if exists) will be skipped, next files will be uploaded event if exists online and have the same size.

Yep, this is part of the previous code logic that I will be focusing on next. You're right, I have that in mind.

> Also doesn't save my profile after exit.

This is weird. Any errors in the log file?

PS: If you want to chat about this more comfortably, send me an email and let me know where i can reach you.

**Comment by st4ck** [comment]
I've sent you an email, by the way in the logs
```
[ 2017-10-19 11:20:51 - Extensions ] : !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! 
[ 2017-10-19 11:20:51 - Extensions ] : Message: Exception has been thrown by the target of an invocation. 
[ 2017-10-19 11:20:51 - Extensions ] : Source: mscorlib Type: System.Reflection.TargetInvocationException 
[ 2017-10-19 11:20:51 - Extensions ] : StackTrace: at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor) at System.Reflection.RuntimeMethodInfo.UnsafeInvokeInternal(Object obj, Object[] parameters, Object[] arguments) at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture) at Newtonsoft.Json.Serialization.JsonContract.<>c__DisplayClass73_0.b__0(Object o, StreamingContext context) at Newtonsoft.Json.Serialization.JsonContract.InvokeOnSerializing(Object o, StreamingContext context) at Newtonsoft.Json.Serialization.JsonSerializerInternalWriter.OnSerializing(JsonWriter writer, JsonContract contract, Object value) at Newtonsoft.Json.Serialization.JsonSerializerInternalWriter.SerializeObject(JsonWriter writer, Object value, JsonObjectContract contract, JsonProperty member, JsonContainerContract collectionContract, JsonProperty containerProperty) at Newtonsoft.Json.Serialization.JsonSerializerInternalWriter.SerializeValue(JsonWriter writer, Object value, JsonContract valueContract, JsonProperty member, JsonContainerContract containerContract, JsonProperty containerProperty) at Newtonsoft.Json.Serialization.JsonSerializerInternalWriter.SerializeList(JsonWriter writer, IEnumerable values, JsonArrayContract contract, JsonProperty member, JsonContainerContract collectionContract, JsonProperty containerProperty) at Newtonsoft.Json.Serialization.JsonSerializerInternalWriter.SerializeValue(JsonWriter writer, Object value, JsonContract valueContract, JsonProperty member, JsonContainerContract containerContract, JsonProperty containerProperty) at Newtonsoft.Json.Serialization.JsonSerializerInternalWriter.Serialize(JsonWriter jsonWriter, Object value, Type objectType) at Newtonsoft.Json.JsonSerializer.SerializeInternal(JsonWriter jsonWriter, Object value, Type objectType) at Newtonsoft.Json.JsonConvert.SerializeObjectInternal(Object value, Type type, JsonSerializer jsonSerializer) at Newtonsoft.Json.JsonConvert.SerializeObject(Object value, Formatting formatting) at FTPboxLib.Settings.SaveProfile(AccountController account) in C:\Users\Roberto\Documents\Visual Studio 2013\Projects\FTPbox-master\FTPboxLib\Settings.cs:line 115 
[ 2017-10-19 11:20:51 - Extensions ] : !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! 
```
During debug VB doesn't throw any errors but It's evident depends on Serialization
